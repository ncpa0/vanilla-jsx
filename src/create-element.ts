import { SignalProxyListenerRef, sigProxy } from "./signals/proxy";

type FunctionComponent = (props: object) => JSX.Element;

type CreateElementProps = {
  [k: string]: any;
  children?: JSX.Children;
};

type ChildNode = Element | DocumentFragment | Text | JSX.Signal<Text | Element>;

function asArray<T>(value: T | T[]): T[] {
  return Array.isArray(value) ? value : [value];
}

function isSignal(o: object): o is JSX.Signal {
  return (
    "add" in o && typeof o["add"] === "function"
  );
}

function valToString(value: JSX.SignalValue): string {
  return value == null ? "" : String(value);
}

const getInitialChild = (appendTo: Element) => {
  const initialChild = document.createTextNode("");
  appendTo.appendChild(initialChild);
  return new WeakRef(initialChild);
};

const childBindingFactory = (lastNodeRef: WeakRef<Text | Element>) => {
  return (_: unknown, value: Text | Element, sigRef?: SignalProxyListenerRef) => {
    const lastNode = lastNodeRef?.deref();
    if (!lastNode) {
      return sigRef?.detach();
    }

    if (typeof value === "string") {
      if (lastNode instanceof Text) {
        lastNode.textContent = valToString(value);
      } else {
        const node = document.createTextNode(valToString(value));
        lastNode.replaceWith(node);
        lastNodeRef = new WeakRef(node);
      }
    } else {
      lastNode.replaceWith(value);
      lastNodeRef = new WeakRef(value);
    }
  };
};

const mapChildren = (
  children: Exclude<JSX.Children, JSX.Element | JSX.VanillaValue>,
  accumulator: Array<ChildNode>,
) => {
  for (let i = 0; i < children.length; i++) {
    const child = children[i];

    switch (typeof child) {
      case "string":
      case "number":
        accumulator.push(document.createTextNode(valToString(child)));
        break;
      case "boolean":
      case "undefined":
        break;
      case "object":
        if (child === null) {
          break;
        }
        if (Array.isArray(child)) {
          mapChildren(child, accumulator);
        } else {
          accumulator.push(child);
        }
        break;
    }
  }
};

const eventBindingFactory = (eventName: string) => {
  let prevListener: (e: Event) => void;
  return (elem: HTMLElement, value: any) => {
    if (prevListener) {
      elem.removeEventListener(eventName, prevListener);
    }
    if (value == null) {
      return;
    }
    elem.addEventListener(eventName, value as any);
    prevListener = value as any;
  };
};

const attributeBindingFactory = (attributeName: string) => {
  return (elem: HTMLElement, value: any) => {
    if (value == null && elem.hasAttribute(attributeName)) {
      elem.removeAttribute(attributeName);
      return;
    }

    if (attributeName in elem) {
      (elem as any)[attributeName] = value;
      return;
    }

    if (value != null) {
      elem.setAttribute(attributeName, valToString(value));
    }
  };
};

export function createElement(
  tag: string | FunctionComponent,
  props: CreateElementProps | undefined,
  ...children: Exclude<JSX.Children, JSX.Element | JSX.VanillaValue>
): JSX.Element {
  if (typeof tag === "function") {
    const forwardedProps = props ?? {};
    if (children.length > 0) {
      forwardedProps.children = children;
    }
    return tag(forwardedProps);
  }

  children = children.length > 0 ? children : props ? asArray(props?.children) : [];

  const childNodes: Array<ChildNode> = [];

  mapChildren(children, childNodes);

  if (tag === "") {
    const fragment = document.createDocumentFragment();
    for (const child of childNodes) {
      if (isSignal(child)) {
        throw new Error("Signals cannot be used as children of fragments.");
      }
      fragment.appendChild(child);
    }
    return fragment as any;
  }

  const element = document.createElement(tag);

  if (props) {
    for (const [propName, propValue] of Object.entries(props)) {
      if (propName === "children") {
        continue;
      }

      if (
        typeof propValue === "object"
        && propValue !== null
        && isSignal(propValue)
      ) {
        const sig = sigProxy(propValue);

        if (propName.startsWith("on")) {
          const eventName = propName.substring(2).toLowerCase();
          sig.bindTo(element, eventBindingFactory(eventName));
        } else {
          sig.bindTo(element, attributeBindingFactory(propName));
        }
      } else {
        if (propName.startsWith("on")) {
          const eventName = propName.substring(2).toLowerCase();
          element.addEventListener(eventName, propValue);
        } else {
          element.setAttribute(propName, propValue);
        }
      }
    }
  }

  for (const child of childNodes) {
    if (isSignal(child)) {
      const sig = sigProxy(child);
      let initialNodeRef: WeakRef<Text | Element> = getInitialChild(element);
      sig.bindTo(element, childBindingFactory(initialNodeRef));
    } else {
      element.appendChild(child);
    }
  }

  return element;
}

const Fragment = "";

export { createElement as jsx, createElement as jsxs, Fragment };
