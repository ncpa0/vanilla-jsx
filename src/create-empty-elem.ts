export function createEmptyElem() {
  const elem = document.createElement("div");
  elem.style.display = "none";
  return elem;
}
