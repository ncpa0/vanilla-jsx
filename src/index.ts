export { $component, type ComponentApi } from "./component";
export { createElement, Fragment, jsx, jsxs } from "./create-element";
export { If, type IfProps } from "./intrinsic/if";
export { Range, type RangeProps } from "./intrinsic/range";
export { Case, type CaseProps, Switch, type SwitchProps } from "./intrinsic/switch";
export { type ClassName } from "./jsx-namespace/jsx.types";
export { bindSignal } from "./signals/proxy";
export {
  deriveMany,
  type DispatchFunc,
  flushBatch,
  type ReadonlySignal,
  setAutoBatchingEnabled,
  sig,
  type Signal,
  signal,
  type SignalListener,
  type SignalListenerReference,
} from "./signals/signal";
