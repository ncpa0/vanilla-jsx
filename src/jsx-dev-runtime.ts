export { Fragment, jsx, jsxs } from "./index";
import { jsx } from "./index";

export function jsxDEV(tag: any, props?: object) {
  let children = [];
  // @ts-expect-error
  if (props && props.children) {
    // @ts-expect-error
    children = Array.isArray(props.children) ? props.children : [props.children];
    // @ts-expect-error
    delete props.children;
  }
  return jsx(tag, props, ...children);
}
