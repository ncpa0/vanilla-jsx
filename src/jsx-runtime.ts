export { Fragment, jsx, jsxs } from "./index";
