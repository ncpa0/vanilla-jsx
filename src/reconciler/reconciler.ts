import { DomInteraction } from "../dom/dom-interaction";
import { InteractionInterface } from "../dom/interaction-interface";
import { SignalProxy, sigProxy } from "../signals/proxy";
import { BindingFactories } from "./factories";
import { asArray, isSignal } from "./utils";

declare global {
  namespace VanillaJSX {
    interface Types {}
  }
}

export type GetElement = VanillaJSX.Types extends
  { Element: infer T extends object } ? T
  : Element;
export type GetTextElement = VanillaJSX.Types extends
  { TextElement: infer T extends object } ? T
  : Text;
export type GetFragmentElement = VanillaJSX.Types extends
  { FragmentElement: infer T extends object } ? T
  : DocumentFragment;
export type GetEvent = VanillaJSX.Types extends { Ev: infer T extends object }
  ? T
  : Event;

type FunctionComponent = (props: object) => JSX.Element;

type CreateElementProps = {
  [k: string]: any;
  children?: JSX.Children;
  unsafeHTML?: boolean;
};

type ChildElement =
  | GetElement
  | GetTextElement
  | GetFragmentElement
  | JSX.Signal;

export class Reconciler {
  private static instance?: Reconciler;

  public static setInteractionInterface(
    interaction: InteractionInterface<
      any,
      any,
      any,
      any
    >,
  ) {
    Reconciler.instance = new Reconciler(interaction);
  }

  public static getInstance() {
    if (!Reconciler.instance) {
      Reconciler.instance = new Reconciler(new DomInteraction());
    }
    return Reconciler.instance;
  }

  public static interactions(): InteractionInterface<
    GetElement,
    GetTextElement,
    GetFragmentElement,
    GetEvent
  > {
    return Reconciler.getInstance().dom;
  }

  private factories;
  constructor(
    private dom: InteractionInterface<
      GetElement,
      GetTextElement,
      GetFragmentElement,
      GetEvent
    >,
  ) {
    this.factories = new BindingFactories(this.dom);
  }

  private appendEmptyTextRef(appendTo: Element) {
    const initialChild = this.dom.createText("");
    this.dom.append(appendTo, initialChild);
    return new WeakRef(initialChild);
  }

  private mapChildren = (
    children: Exclude<JSX.Children, JSX.Element | JSX.VanillaValue>,
    accumulator: Array<ChildElement>,
    unsafe?: boolean,
  ) => {
    for (let i = 0; i < children.length; i++) {
      const child = children[i];

      switch (typeof child) {
        case "string":
          if (unsafe) {
            const nodes = this.dom.parseUnsafe(child);
            for (const node of nodes) {
              accumulator.push(node);
            }
          } else {
            accumulator.push(this.dom.createText(child));
          }
          break;
        case "number":
        case "bigint":
          accumulator.push(this.dom.createText(child));
        case "boolean":
        case "undefined":
          break;
        case "object":
          if (child === null) {
            break;
          }
          if (Array.isArray(child)) {
            this.mapChildren(child, accumulator);
          } else {
            accumulator.push(child);
          }
          break;
      }
    }
  };

  public createElement(
    tag: string | FunctionComponent,
    props: CreateElementProps | undefined,
    ...children: Exclude<JSX.Children, JSX.Element | JSX.VanillaValue>
  ) {
    if (typeof tag === "function") {
      const forwardedProps = props ?? {};
      if (children.length > 0) {
        forwardedProps.children = children;
      }
      return tag(forwardedProps);
    }

    children = children.length > 0
      ? children
      : props
      ? asArray(props?.children)
      : [];

    const childNodes: Array<ChildElement> = [];

    this.mapChildren(children, childNodes, props?.unsafeHTML);

    if (tag === "") {
      const fragment = this.dom.createFragment();
      for (const child of childNodes) {
        if (isSignal(child)) {
          throw new Error("Signals cannot be used as children of fragments.");
        }
        this.dom.append(fragment, child);
      }
      return fragment as any;
    }

    const element = this.dom.create(tag);

    if (props) {
      for (const [propName, propValue] of Object.entries(props)) {
        if (propName === "children" || propName === "unsafeHTML") {
          continue;
        }

        if (
          typeof propValue === "object"
          && propValue !== null
          && isSignal(propValue)
        ) {
          const sig = sigProxy(propValue);

          if (propName.startsWith("on")) {
            const eventName = propName.substring(2).toLowerCase();
            sig.bindTo(element, this.factories.createEventBinding(eventName));
          } else {
            sig.bindTo(
              element,
              this.factories.createAttributeBinding(propName),
            );
          }
        } else {
          if (propName.startsWith("on")) {
            const eventName = propName.substring(2).toLowerCase();
            this.factories.createEventBinding(eventName)(element, propValue);
          } else {
            this.factories.createAttributeBinding(propName)(element, propValue);
          }
        }
      }
    }

    for (const child of childNodes) {
      if (isSignal(child)) {
        const sig = sigProxy(child) as SignalProxy<Element | Text | undefined>;
        let initialNodeRef = this.appendEmptyTextRef(element);
        sig.bindTo(element, this.factories.createChildBinding(initialNodeRef));
      } else {
        this.dom.append(element, child);
      }
    }

    return element;
  }
}

export const createElement: Reconciler["createElement"] = (...args) =>
  Reconciler.getInstance().createElement(...args);

export const setInteractionInterface:
  (typeof Reconciler)["setInteractionInterface"] = (arg) =>
    Reconciler.setInteractionInterface(arg);

const Fragment = "";

export { createElement as jsx, createElement as jsxs, Fragment };
