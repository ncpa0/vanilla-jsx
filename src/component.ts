import { createElement } from "./reconciler/reconciler";
import { SignalProxy, sigProxy } from "./sig-proxy/_proxy";

declare global {
  namespace JSX {
    interface IntrinsicElements {
      "wc-wrapper": HTMLProps<VanillaJSX.DivTagProps>;
    }
  }
}

class ComponentNode extends HTMLElement {
  static readonly TAG_NAME = "vjsx-component-node";
  static readonly MOUNTED = "mounted";
  static readonly UNMOUNTED = "unmounted";
  readonly MOUNTED = ComponentNode.MOUNTED;
  readonly UNMOUNTED = ComponentNode.UNMOUNTED;

  _emitter = new EventTarget();

  connectedCallback() {
    this.style.display = "contents";
    this._emitter.dispatchEvent(
      new CustomEvent(ComponentNode.MOUNTED, { detail: { target: this } }),
    );
  }

  disconnectedCallback() {
    this._emitter.dispatchEvent(
      new CustomEvent(ComponentNode.UNMOUNTED, { detail: { target: this } }),
    );
  }
}

if (typeof window !== "undefined") {
  window.customElements.define(ComponentNode.TAG_NAME, ComponentNode);
}

export interface OnChangeCallback {
  (): void | (() => void);
}

export interface OnMountCallback {
  (): void | (() => void);
}

export interface OnUnmountCallback {
  (): void;
}

export interface ComponentApi {
  /**
   * Register a callback to be called when the component is mounted into the
   * document. This can be called multiple times if the component is
   * unmounted and then re-mounted.
   */
  onMount(listener: OnMountCallback): void;
  /**
   * Register a callback to be called when the component is unmounted from the
   * document. This can be called multiple times if the component is
   * re-mounted and then unmounted again.
   */
  onUnmount(listener: OnUnmountCallback): void;
  /**
   * Register a callback to be called when the given dependencies change.
   * This callback will only be invoked while the component is mounted.
   */
  onChange(cb: OnChangeCallback, deps: Array<JSX.Signal>): void;
}

const noop = () => {};

class CAPI implements ComponentApi {
  private isConnected = false;
  private changeHandlers: Array<
    [cb: OnChangeCallback, deps: JSX.Signal<any>[]]
  > = [];
  private mountHandlers: OnMountCallback[] = [];
  private unmountHandlers: OnUnmountCallback[] = [];

  private connectMountHandlers(element: ComponentNode) {
    for (let i = 0; i < this.mountHandlers.length; i++) {
      const handler = this.mountHandlers[i]!;
      element._emitter.addEventListener(ComponentNode.MOUNTED, () => {
        const cleanup = handler();
        if (cleanup) {
          element._emitter.addEventListener(
            ComponentNode.UNMOUNTED,
            cleanup,
            { once: true },
          );
        }
      });
    }
    this.mountHandlers.splice(0, this.mountHandlers.length);
  }

  private connectUnmountHandlers(element: ComponentNode) {
    for (let i = 0; i < this.unmountHandlers.length; i++) {
      const handler = this.unmountHandlers[i]!;
      element._emitter.addEventListener(ComponentNode.UNMOUNTED, handler);
    }
    this.unmountHandlers.splice(0, this.unmountHandlers.length);
  }

  private connectChangeHandlers(element: ComponentNode) {
    for (let i = 0; i < this.changeHandlers.length; i++) {
      const [cb, deps] = this.changeHandlers[i]!;
      let isQueued = false;
      let cleanup: () => void = noop;

      const addBinding = (signal: SignalProxy<any>, elem: ComponentNode) => {
        const unbind = signal.add(() => {
          if (isQueued) return;
          isQueued = true;
          queueMicrotask(() => {
            isQueued = false;
            cleanup();
            cleanup = cb() ?? noop;
          });
        });

        elem._emitter.addEventListener(ComponentNode.UNMOUNTED, () => {
          unbind();
          cleanup();
        }, {
          once: true,
        });
      };

      for (let j = 0; j < deps.length; j++) {
        const signal = sigProxy(deps[j]!);

        element._emitter.addEventListener(ComponentNode.MOUNTED, (ev) => {
          const event = ev as CustomEvent<{ target: ComponentNode }>;
          addBinding(signal, event.detail.target);
        });
      }
    }
    this.changeHandlers.splice(0, this.changeHandlers.length);
  }

  onChange(...args: [cb: OnChangeCallback, deps: JSX.Signal<any>[]]): void {
    if (this.isConnected) {
      throw new Error(
        "Cannot register onChange handler after component initialization.",
      );
    }
    this.changeHandlers.push(args);
  }

  onMount(listener: OnMountCallback): void {
    if (this.isConnected) {
      throw new Error(
        "Cannot register onMount handler after component initialization.",
      );
    }
    this.mountHandlers.push(listener);
  }

  onUnmount(listener: OnUnmountCallback): void {
    if (this.isConnected) {
      throw new Error(
        "Cannot register onUnmount handler after component initialization.",
      );
    }
    this.unmountHandlers.push(listener);
  }

  connect(element: ComponentNode) {
    this.connectMountHandlers(element);
    this.connectUnmountHandlers(element);
    this.connectChangeHandlers(element);
    this.isConnected = true;
  }
}

/**
 * Wraps the returned html elements in a Web Component and provides an API
 * for subscribing to that component's lifecycle events.
 *
 * @example
 * const MyComponent = $component<{foo: JSX.Signal<string>}>((props, api) => {
 *   api.onMount(() => {
 *     console.log("Mounted");
 *   });
 *   api.onUnmount(() => {
 *     console.log("Unmounted");
 *   });
 *   api.onChange(() => {
 *     console.log("Foo changed");
 *   }, [props.foo]);
 *
 *   return <div>{props.foo}</div>;
 * });
 */
export const $component = <P>(
  component: (props: P, api: ComponentApi) => JSX.Element,
) => {
  return (props: P) => {
    const api = new CAPI();
    const elem = createElement(
      ComponentNode.TAG_NAME,
      {},
      component(props, api),
    ) as ComponentNode;
    api.connect(elem);
    return elem;
  };
};
