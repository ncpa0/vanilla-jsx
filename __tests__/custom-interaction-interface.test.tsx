import { beforeAll, describe, expect, it } from "vitest";
import {
  InteractionInterface,
  Primitive,
} from "../src/dom/interaction-interface";
import { Fragment, jsx } from "../src/index";
import { setInteractionInterface } from "../src/reconciler/reconciler";

class CustomFragment {
  parent?: CustomElement | CustomFragment;
  children: Array<CustomElement | CustomTextNode | CustomFragment> = [];

  append(child: CustomElement | CustomTextNode | CustomFragment) {
    this.children.push(child);
    child.parent = this;
  }

  replaceChild(
    newChild: CustomElement | CustomTextNode | CustomFragment,
    oldChild: CustomElement | CustomTextNode | CustomFragment,
  ) {
    const index = this.children.indexOf(oldChild);
    if (index !== -1) {
      this.children[index]!.parent = undefined;
      this.children[index] = newChild;
      newChild.parent = this;
    }
  }
}

class CustomElement extends CustomFragment {
  attributes: Array<[string, string]> = [];
  listeners: Array<[string, (event: CustomEvent) => void]> = [];
  dataset: Record<string, string> = {};
  classNames = new Set<string>();
  isHidden = false;
  style: Record<string, string> = {};

  constructor(public readonly tagName: string) {
    super();
  }

  setAttribute(name: string, value: string) {
    this.attributes.push([name, value]);
  }

  removeAttribute(name: string) {
    this.attributes = this.attributes.filter(([key]) => key !== name);
  }

  addEventListener(name: string, listener: (event: CustomEvent) => void) {
    this.listeners.push([name, listener]);
  }

  removeEventListener(name: string, listener: (event: CustomEvent) => void) {
    this.listeners = this.listeners.filter(
      ([key, value]) => key !== name || value !== listener,
    );
  }

  setData(key: string, value: string) {
    this.dataset[key] = value;
  }

  deleteData(key: string) {
    delete this.dataset[key];
  }

  addClassNames(classNames: string[]) {
    for (const className of classNames) {
      this.classNames.add(className);
    }
  }

  removeClassNames(classNames: string[]) {
    for (const className of classNames) {
      this.classNames.delete(className);
    }
  }

  setClassName(value: string) {
    this.classNames = new Set(value.split(" ").filter(Boolean));
  }
}

class CustomTextNode {
  parent?: CustomElement | CustomFragment;
  constructor(public content: string) {}
}

class CustomEvent {}

class CustomInteractions
  implements
    InteractionInterface<
      CustomElement,
      CustomTextNode,
      CustomFragment,
      CustomEvent
    >
{
  create(tag: string): CustomElement {
    return new CustomElement(tag);
  }

  createText(content: Primitive): CustomTextNode {
    return new CustomTextNode(String(content));
  }

  createFragment(): CustomFragment {
    return new CustomFragment();
  }

  append(
    parent: CustomFragment | CustomElement,
    child: CustomFragment | CustomElement | CustomTextNode,
  ): void {
    parent.append(child);
  }

  remove(child: CustomElement | CustomTextNode): void {
    const parent = child.parent;
    if (parent) {
      parent.replaceChild(this.createFragment(), child);
    }
  }

  replace(
    oldChild: CustomElement | CustomTextNode,
    newChild: CustomElement | CustomFragment | CustomTextNode,
  ): void {
    const parent = oldChild.parent;
    if (parent) {
      parent.replaceChild(newChild, oldChild);
    }
  }

  setText(element: CustomTextNode, value: Primitive): void {
    element.content = String(value);
  }

  setClass(element: CustomElement, value: string): void {
    element.setClassName(value);
  }

  addClassName(element: CustomElement, ...value: string[]): void {
    element.addClassNames(value);
  }

  removeClassName(element: CustomElement, ...value: string[]): void {
    element.removeClassNames(value);
  }

  setAttributeOrProperty(
    element: CustomElement,
    name: string,
    value: Primitive,
  ): void {
    this.setAttribute(element, name, value);
  }

  setProperty(element: CustomElement, name: string, value: Primitive): void {
    this.setAttribute(element, name, value);
  }

  setAttribute(element: CustomElement, name: string, value: Primitive): void {
    if (value == null) {
      element.removeAttribute(name);
      return;
    }
    element.setAttribute(name, String(value));
  }

  setData(element: CustomElement, name: string, value: Primitive): void {
    if (value == null) {
      element.deleteData(name);
      return;
    }
    element.setData(name, String(value));
  }

  on(
    element: CustomElement,
    event: string,
    listener: (event: CustomEvent) => void,
  ): void {
    element.addEventListener(event, listener);
  }

  once(
    element: CustomElement,
    event: string,
    listener: (event: CustomEvent) => void,
  ): void {
    element.addEventListener(event, listener);
  }

  off(
    element: CustomElement,
    event: string,
    listener: (event: CustomEvent) => void,
  ): void {
    element.removeEventListener(event, listener);
  }

  parseUnsafe(
    content: string,
  ): Iterable<CustomElement | CustomTextNode | CustomFragment> {
    return [];
  }

  isText(
    element: CustomTextNode | CustomElement | CustomFragment,
  ): element is CustomTextNode {
    return element instanceof CustomTextNode;
  }

  isFragment(
    element: CustomElement | CustomFragment | CustomTextNode,
  ): element is CustomFragment {
    return element instanceof CustomFragment;
  }

  hide(element: CustomElement): void {
    element.isHidden = true;
  }

  replaceAllChildren(
    parent: CustomElement | CustomFragment,
    ...children: (CustomElement | CustomFragment | CustomTextNode)[]
  ): void {
    for (const child of parent.children) {
      child.parent = undefined;
    }
    parent.children = children;
    for (const child of children) {
      child.parent = parent;
    }
  }

  insertBefore(
    parent: CustomElement | CustomFragment,
    child: CustomElement | CustomTextNode | CustomFragment,
    before: CustomElement | CustomTextNode | CustomFragment,
  ): void {
    const index = parent.children.indexOf(before);
    if (index !== -1) {
      parent.children.splice(index, 0, child);
      child.parent = parent;
    }
  }

  clearStyle(element: CustomElement): void {
    element.style = {};
  }

  insertAfter(
    child: CustomElement | CustomTextNode | CustomFragment,
    after: CustomElement | CustomTextNode | CustomFragment,
  ): void {
    const parent = after.parent;
    if (parent) {
      const index = parent.children.indexOf(after);
      if (index !== -1) {
        parent.children.splice(index + 1, 0, child);
        child.parent = parent;
      }
    }
  }

  setStyle(
    element: CustomElement,
    styleKey: string,
    value: string | undefined,
  ): void {
    if (value == null) {
      delete element.style[styleKey];
      return;
    }
    element.style[styleKey] = value;
  }
}

describe("correctly works with a custom InteractionInterface", () => {
  beforeAll(() => {
    setInteractionInterface(new CustomInteractions());
  });

  it("should render a simple element", () => {
    const element = <div />;
    expect(element).toEqual(new CustomElement("div"));
    expect(element).toMatchSnapshot();
  });

  it("should render a more complex element structure", () => {
    const element = (
      <div id="root" draggable>
        <p class="header text">Hello</p>
        <>
          <span data-foo="1">World</span>
        </>
      </div>
    );

    const root = new CustomElement("div");
    root.attributes = [
      ["id", "root"],
      ["draggable", "true"],
    ];
    const paragraph = new CustomElement("p");
    paragraph.classNames = new Set(["header", "text"]);
    paragraph.append(new CustomTextNode("Hello"));
    root.append(paragraph);
    const fragment = new CustomFragment();
    const span = new CustomElement("span");
    span.dataset = { foo: "1" };
    span.append(new CustomTextNode("World"));
    fragment.append(span);
    root.append(fragment);

    expect(element).toEqual(root);
    expect(element).toMatchSnapshot();
  });
});
