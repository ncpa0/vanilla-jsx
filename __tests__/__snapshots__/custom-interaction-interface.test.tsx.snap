// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`correctly works with a custom InteractionInterface > should render a more complex element structure 1`] = `
CustomElement {
  "attributes": [
    [
      "id",
      "root",
    ],
    [
      "draggable",
      "true",
    ],
  ],
  "children": [
    CustomElement {
      "attributes": [],
      "children": [
        CustomTextNode {
          "content": "Hello",
          "parent": [Circular],
        },
      ],
      "classNames": Set {
        "header",
        "text",
      },
      "dataset": {},
      "isHidden": false,
      "listeners": [],
      "parent": [Circular],
      "tagName": "p",
    },
    CustomFragment {
      "children": [
        CustomElement {
          "attributes": [],
          "children": [
            CustomTextNode {
              "content": "World",
              "parent": [Circular],
            },
          ],
          "classNames": Set {},
          "dataset": {
            "foo": "1",
          },
          "isHidden": false,
          "listeners": [],
          "parent": [Circular],
          "tagName": "span",
        },
      ],
      "parent": [Circular],
    },
  ],
  "classNames": Set {},
  "dataset": {},
  "isHidden": false,
  "listeners": [],
  "parent": undefined,
  "tagName": "div",
}
`;

exports[`correctly works with a custom InteractionInterface > should render a simple element 1`] = `
CustomElement {
  "attributes": [],
  "children": [],
  "classNames": Set {},
  "dataset": {},
  "isHidden": false,
  "listeners": [],
  "parent": undefined,
  "tagName": "div",
}
`;
